net: dhcpv4_server: Improve declined addresses management

rlubos opened the pull request:
In case conflict is detected (either due to receiving Decline message or
due to ICMP probe getting reply), the conflicting address becomes
blocked for further use.

Although the RFC is not specific about how long should the address be
blocked, it make sense to implement some fallback mechanisms to reuse
blocked addresses in the server, otherwise, after longer period of
operation, it may run out of usable address.

This commit adds a timeout for declined addresses, so that by default
the address is marked back as "free" after 24 hrs (default lease time).
It also implements a mechanism, which allows to re-use the oldest
declined entry in case the server runs out of fresh addresses to assign.

rlubos committed:
net: dhcpv4_server: Improve declined addresses management

In case conflict is detected (either due to receiving Decline message or
due to ICMP probe getting reply), the conflicting address becomes
blocked for further use.

Although the RFC is not specific about how long should the address be
blocked, it make sense to implement some fallback mechanisms to reuse
blocked addresses in the server, otherwise, after longer period of
operation, it may run out of usable address.

This commit adds a timeout for declined addresses, so that by default
the address is marked back as "free" after 24 hrs (default lease time).
It also implements a mechanism, which allows to re-use the oldest
declined entry in case the server runs out of fresh addresses to assign.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: dhcpv4_server: Add tests for declined address expiry/reuse

Verify that declined addresses are not permanently blocked - i. e.
expire after configured timeout and in case of free address shortage,
can be reused.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar commented on subsys/net/lib/dhcpv4/Kconfig:
```diff
@@ -85,6 +85,17 @@ config NET_DHCPV4_SERVER_ADDR_LEASE_TIME
+config NET_DHCPV4_SERVER_ADDR_DECLINE_TIME
+	int "The time IPv4 addresses remains blocked after conflict detection (seconds)"
+	default 86400
+	help
```
Would it make sense to add a range for allowable values?
We do not have ranges for other int values in this kconfig either which might be a bit problematic especially as we do not check the value in the code.

rlubos replied:
I just wonder what would be correct range values, it's not specified by RFC so we'd need to choose some arbitrary ones. That's why I've left it w/o range.

jukkar replied:
I was mostly concerned if user sets timeout as negative value, and that value is fed directly to `K_SECONDS()`, what is the timeout value in that case. This is of course quite academic, why would anyone put <0 value here.
Anyway, I was just thinking what could go wrong. We could limit the value to [0, MAX_INT] for example.

rlubos replied:
Good point with the negative values. Ok, I've applied ranges for the time options, for lease (and decline) time uint32_t max value, as it's the theoretical maximum that lease time option can carry. For probe timeout, 60 seconds max, the maximum is already high enough to make the server replies sluggish, I don't think setting anything higher would make sense at all.

jukkar reviewed (commented):
Overall LGTM, minor nit about allowed values of the Kconfig option(s).

rlubos committed:
net: dhcpv4_server: Apply range limits for Kconfig options

Apply ranges to DHCPv4 server timeout Kconfig options, so that it cannot
be set to a negative value by mistake.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar reviewed (approved):
Thanks, LGTM
